server/src/test/java/org/elasticsearch/cluster/coordination/ElasticsearchNodeCommandIT.java

```diff
@@ -418,4 +423,57 @@ public void testCanRunUnsafeBootstrapAfterErroneousDetachWithoutLoosingMetaData(
         assertThat(state.metaData().settings().get(INDICES_RECOVERY_MAX_BYTES_PER_SEC_SETTING.getKey()),
                 equalTo("1234kb"));
     }
+
+    private static class SimulatedDeleteFailureException extends RuntimeException {
+    }
+
+    public void testCleanupOldMetaDataFails() throws Exception {
+        // establish some metadata.
+        internalCluster().setBootstrapMasterNodeIndex(0);
+        internalCluster().startNode();
+        Environment environment = TestEnvironment.newEnvironment(internalCluster().getDefaultSettings());
+        internalCluster().stopRandomDataNode();
+
+        // find data paths
+        Path[] dataPaths;
+        try (NodeEnvironment nodeEnvironment = new NodeEnvironment(environment.settings(), environment)) {
+            dataPaths = nodeEnvironment.nodeDataPaths();
+        }
+
+        NamedXContentRegistry namedXContentRegistry = new NamedXContentRegistry(ClusterModule.getNamedXWriteables());
+
+        final Manifest originalManifest = loadLatestManifest(dataPaths, namedXContentRegistry);
+        final MetaData originalMetaData = loadMetaData(dataPaths, namedXContentRegistry, originalManifest);
+
+        executeCommand(new UnsafeBootstrapMasterCommand() {
+            @Override
+            protected void cleanUpOldMetaData(Terminal terminal, Path[] dataPaths, long newGeneration) {
+                throw new SimulatedDeleteFailureException();
+            }
+        }, environment, 0, false);
+
+
+        // check original meta-data left untouched.
+        assertEquals(loadMetaData(dataPaths, namedXContentRegistry, originalManifest).clusterUUID(), originalMetaData.clusterUUID());
+
+        // check that we got new clusterUUID despite deletion failing
+        final Manifest secondManifest = loadLatestManifest(dataPaths, namedXContentRegistry);
+        final MetaData secondMetaData = loadMetaData(dataPaths, namedXContentRegistry, secondManifest);
+        assertThat(secondManifest.getGlobalGeneration(), greaterThan(originalManifest.getGlobalGeneration()));
+        assertNotEquals(originalMetaData.clusterUUID(), secondMetaData.clusterUUID());
+
+        // check that a new run will cleanup.
+        executeCommand(new UnsafeBootstrapMasterCommand(), environment, 0, false);
+
+        assertNull(loadMetaData(dataPaths, namedXContentRegistry, originalManifest));
+        assertNull(loadMetaData(dataPaths, namedXContentRegistry, secondManifest));
```

Comment:
Can we also check that loadLatestManifest and loadMetadata will load non-null objects?

Reply:
Fixed in ca59377ea745834b2d1477a1cde5d62cfc68fe77